src/main/kotlin/org/rust/ide/commenter/RsCommenter.kt

```diff
@@ -27,8 +44,84 @@ class RsCommenter : Commenter, CodeDocumentationAwareCommenter {
     override fun getBlockCommentPrefix(): String = "/*"
     override fun getBlockCommentSuffix(): String = "*/"
 
-    // for nested comments
-
+    // unused because we implement SelfManagingCommenter
     override fun getCommentedBlockCommentPrefix(): String = "*//*"
     override fun getCommentedBlockCommentSuffix(): String = "*//*"
+
+    override fun getBlockCommentPrefix(selectionStart: Int, document: Document, data: CommentHolder): String? =
+        blockCommentPrefix
+
+    override fun getBlockCommentSuffix(selectionEnd: Int, document: Document, data: CommentHolder): String? =
+        blockCommentSuffix
+
+    override fun getBlockCommentRange(
+        selectionStart: Int,
+        selectionEnd: Int,
+        document: Document,
+        data: CommentHolder
+    ): TextRange? =
+        SelfManagingCommenterUtil.getBlockCommentRange(selectionStart, selectionEnd, document, blockCommentPrefix, blockCommentSuffix)
+
+    override fun insertBlockComment(
+        startOffset: Int,
+        endOffset: Int,
+        document: Document,
+        data: CommentHolder?
+    ): TextRange =
+        SelfManagingCommenterUtil.insertBlockComment(startOffset, endOffset, document, blockCommentPrefix, blockCommentSuffix)
+
+    override fun uncommentBlockComment(
+        startOffset: Int,
+        endOffset: Int,
+        document: Document,
+        data: CommentHolder?
+    ) =
+        SelfManagingCommenterUtil.uncommentBlockComment(startOffset, endOffset, document, blockCommentPrefix, blockCommentSuffix)
+
+    override fun isLineCommented(line: Int, offset: Int, document: Document, data: CommentHolder): Boolean {
+        return getStartLineComment(line, document, data.file)?.isEolComment ?: false
```

Comment:
Hmm, this can be implemented with `CharArrayUtil.regionMatches` too

Reply:
I thought that I would have to match three separate strings (`//`, `///`, `//!`), but actually it would be enough to check for `//`, right? I can change it if you want. Now I also realized that the order of `LINE_PREFIXES` is important for uncomment to work, so it maybe warrants a comment.